rs/pocket_ic_server/src/pocket_ic.rs

```diff
@@ -831,6 +843,233 @@ impl Operation for GetTime {
+                subnet_id: subnet.get_subnet_id().get().0,
+                request_id: id.get(),
+                http_method: http_method_from(&c.http_method),
+                url: c.url,
+                headers: c.headers.iter().map(http_header_from).collect(),
+                body: c.body,
+                max_response_bytes: c.max_response_bytes.map(|b| b.get()),
+            })
+            .collect();
+        canister_http.append(&mut cur);
+    }
+    canister_http
+}
+
+impl Operation for GetCanisterHttp {
+    fn compute(&self, pic: &mut PocketIc) -> OpOut {
+        let canister_http_requests = get_canister_http_requests(pic);
+        OpOut::CanisterHttp(canister_http_requests)
+    }
+
+    fn id(&self) -> OpId {
+        OpId("get_canister_http".into())
+    }
+}
+
+// START COPY from rs/https_outcalls/client/src/client.rs
+
+#[derive(Clone)]
+pub struct SingleResponseAdapter {
+    response: Result<CanisterHttpSendResponse, (Code, String)>,
+}
+
+impl SingleResponseAdapter {
+    fn new(response: Result<CanisterHttpSendResponse, (Code, String)>) -> Self {
+        Self { response }
+    }
+}
+
+#[tonic::async_trait]
+impl CanisterHttpService for SingleResponseAdapter {
+    async fn canister_http_send(
+        &self,
+        _request: Request<CanisterHttpSendRequest>,
+    ) -> Result<Response<CanisterHttpSendResponse>, Status> {
+        match self.response.clone() {
+            Ok(resp) => Ok(Response::new(resp)),
+            Err((code, msg)) => Err(Status::new(code, msg)),
+        }
+    }
+}
+
+async fn setup_adapter_mock(
+    adapter_response: Result<CanisterHttpSendResponse, (Code, String)>,
+) -> Channel {
+    let (client, server) = tokio::io::duplex(1024);
+    let mock_adapter = SingleResponseAdapter::new(adapter_response);
+    tokio::spawn(async move {
+        Server::builder()
+            .add_service(CanisterHttpServiceServer::new(mock_adapter))
+            .serve_with_incoming(futures::stream::iter(vec![Ok::<_, std::io::Error>(server)]))
+            .await
+    });
+
+    let mut client = Some(client);
+    Endpoint::try_from("http://[::]:50051")
+        .unwrap()
+        .connect_with_connector(service_fn(move |_: Uri| {
+            let client = client.take();
+
+            async move {
+                if let Some(client) = client {
+                    Ok(client)
+                } else {
+                    Err(std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        "Client already taken",
+                    ))
+                }
+            }
+        }))
+        .await
+        .unwrap()
+}
+
+// END COPY
+
+fn process_mock_canister_https_response(
+    pic: &PocketIc,
+    mock_canister_http_response: &MockCanisterHttpResponse,
+) -> OpOut {
+    let subnet_id =
+        ic_types::SubnetId::new(ic_types::PrincipalId(mock_canister_http_response.subnet_id));
+    let subnet = pic.get_subnet_with_id(subnet_id);
+    match subnet {
```

Comment:
nit: These match statements are a bit hard to read because of the nesting. Maybe we can factor some things out into private functions?

Reply:
I've also dropped this match. Is this better now or do you suggest to go further?

Reply:
Much better now, thanks!